feat(relation): migration export

Aflynn50 opened the pull request:
Implement relation migration export.

Note this cannot be tested yet as relation import is not working yet.

This PR includes a WIP for relation import, it will be rebased once it has been fully implemented.
## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps
Unit tests (for now).

## Links

**Jira card:** [JUJU-7354](https://warthogs.atlassian.net/browse/JUJU-7354)

gfouillet commented on domain/application/state/application_endpoint.go:
```diff
@@ -55,6 +55,7 @@ func (st *State) insertApplicationEndpoints(ctx context.Context, tx *sqlair.TX,
 		return internalerrors.Errorf("getting charm relation names: %w", err)
 	}
 
+	st.logger.Criticalf(ctx, "insertApplicationEndpoints(%+v)", relations)
```
**todo:** remove this log.

gfouillet commented on domain/application/state/peer_relation.go:
```diff
@@ -56,13 +56,14 @@ func (st *State) insertPeerRelations(ctx context.Context, tx *sqlair.TX, appUUID
 		return errors.Errorf("getting peer endpoints: %w", err)
 	}
-
+	st.logger.Criticalf(ctx, "insertPeerRelations(%+v)", peerEndpoints)
```
**todo:** remove this log.

gfouillet commented on domain/application/state/peer_relation.go:
```diff
@@ -56,13 +56,14 @@ func (st *State) insertPeerRelations(ctx context.Context, tx *sqlair.TX, appUUID
 		if err != nil {
 			return errors.Errorf("inserting new relation for peer endpoint %q: %w", peer.Name, err)
 		}
+		st.logger.Criticalf(ctx, "insertPeerRelations(%+v) %q", peer, relUUID)
```
**todo:** remove this log.

gfouillet commented on domain/relation/service/relation.go:
```diff
@@ -962,3 +987,120 @@ func (s *Service) SetRelationUnitSettings(
+// relations to insert from the arguments, then inserts them at the end so as to
+// wait as long as possible before turning into a write transaction.
+func (s *Service) ImportRelations(ctx context.Context, args relation.ImportRelationsArgs) error {
+	s.logger.Criticalf(ctx, "ImportRelations(%+v)", args)
```
**todo:** remove this log.

gfouillet commented on domain/relation/service/relation.go:
```diff
@@ -962,3 +987,120 @@ func (s *Service) SetRelationUnitSettings(
+	var relUUID corerelation.UUID
+
+	eps := arg.Key.EndpointIdentifiers()
+	s.logger.Criticalf(ctx, "importRelation(%d): %+v", arg.ID, eps)
```
**todo:** remove this log.

Aflynn50 replied:
These are all from the WIP PR this is based on, they'll go when I rebase it!

gfouillet replied:
Ah, ok sorry.

gfouillet reviewed (commented):
First pass: remove this log ;)

hmlanigan commented on domain/application/state/application.go:
```diff
@@ -90,6 +90,7 @@ func (st *State) CreateApplication(
 	args application.AddApplicationArg,
 	units []application.AddUnitArg,
 ) (coreapplication.ID, error) {
+	st.logger.Criticalf(ctx, "CreateApplication(%q)", name)
```
The critical logging messages are part of debugging the migration import code and unrelated to this PR.

SimonRichardson commented on domain/relation/service/relation.go:
```diff
@@ -962,3 +987,120 @@ func (s *Service) SetRelationUnitSettings(
+
+func settingsMap(in map[string]interface{}) map[string]string {
+	return transform.Map(in, func(k string, v interface{}) (string, string) {
+		return k, fmt.Sprintf("%v", v)
```
Don't do this.

Switch on the types you do expect and convert them to a string. There could be subtle differences between different go versions.

Aflynn50 replied:
This is from https://github.com/juju/juju/pull/19597 which this PR is rebased on.

hmlanigan reviewed (commented):
A few comments.  We need many PRs to land to properly test this, including reverting a644ea469fb735f09dd1159c7e856d055947237c .

hmlanigan commented:
@Aflynn50 I added another sha and got setup which I could almost test export with. However we need different charms to use, as `juju-qa-dummy-source` and `juju-qa-dummy-sink` require a shared token to become idle. For some reason adding the token to config isn't making the applications happy, thus I cannot migrate the model.

gfouillet commented on domain/relation/state/types.go:
```diff
@@ -205,10 +205,18 @@ func (g goalStateData) convertToGoalStateRelationData() relation.GoalStateRelati
-	// EndpointUUID is a unique identifier for the application endpoint
-	EndpointUUID corerelation.EndpointUUID `db:"endpoint_uuid"`
+// exportEndpoint contains information needed to export a relation endpoint.
+type exportEndpoint struct {
```
**thought:** I believe this difference has been made for the wrong reason. The core issue is that we have mixed `relation_endpoint` and `application_endpoint` generally in the code.

Aflynn50 replied:
Yea we have, I'm going to put this on the list of things we can fixup later

gfouillet commented on domain/relation/state/relation.go:
```diff
@@ -2833,11 +2863,134 @@ DELETE FROM relation_endpoint
+	return exportRelations, nil
+}
+
+type relationUnitUUIDAndName struct {
```
**todo:** move it in `types.go`, or reuse  `getRelationUnit` if possible

gfouillet commented on domain/schema/model/sql/0024-relation.sql:
```diff
@@ -187,7 +187,7 @@ INSERT INTO relation_status_type VALUES
 CREATE VIEW v_application_endpoint AS
 SELECT
-    ae.uuid AS endpoint_uuid,
+    ae.uuid AS application_endpoint_uuid,
```
**suggestion** since this is `v_application_endpoint`... This field should be `uuid`...

Aflynn50 replied:
I'm going to leave it as application_endpoint_uuid, just because there are too many places where we've got those  confused, and its good to disambiguate earlier rather than making people check the view.

gfouillet commented on domain/schema/model/sql/0024-relation.sql:
```diff
@@ -213,7 +213,8 @@ JOIN charm_relation AS cr ON ae.charm_relation_uuid = cr.uuid;
 CREATE VIEW v_relation_endpoint AS
 SELECT
-    re.endpoint_uuid,
+    re.uuid AS relation_endpoint_uuid,
```
**suggestion** since this is `v_relation_endpoint`... This field should be `uuid`...

gfouillet reviewed (commented):
I merged delete state and watcher stuff... with `git revert a644ea469fb735f09dd1159c7e856d055947237c` we should be close to be able to QA something ?

Maybe we should add some QA step (even with warning indicating that we may not be to perform it)

Aflynn50 commented on domain/relation/types.go:
```diff
@@ -266,3 +266,39 @@ type GoalStateRelationData struct {
+type ExportEndpoint struct {
+	ApplicationName     string
+	Name                string
+	Role                charm.RelationRole
+	Interface           string
+	Optional            bool
+	Limit               int
+	Scope               charm.RelationScope
```
I exported these in the name of generally being conservative with migration, and the fact that the description package is used for other things as well. We decided for the application endpoints to export the data even though it wasn't needed for import. Do we want to do it differently here?

Aflynn50 commented on domain/relation/state/types.go:
```diff
@@ -205,10 +205,18 @@ func (g goalStateData) convertToGoalStateRelationData() relation.GoalStateRelati
+
+// Endpoint is used to fetch an endpoint from the database. Endpoint is a public
+// struct to allow for embedding in exportEndpoint.
+type Endpoint struct {
```
Because otherwise SQLair can't see its fields when its embedded, it can only see and set exported embedded structs.

gfouillet reviewed (approved):
LGTM

Aflynn50 commented:
> There a bug in here, the description model isn't getting the unit settings. The error is from the description package `Validate()` method.
> 
> ```
> $ juju migrate seven dst
> ERROR missing relation settings for units [sink/0] in relation 0
> 
> > select * from relation_unit
> uuid					relation_endpoint_uuid			unit_uuid				departing
> e55c3a39-6309-4a6e-8f07-da8b3a2f553f	daad9426-5b28-43dc-882c-06fcbc8f3faf	d0c384e3-6800-4b3c-800d-1a37740aa189	false
> ba4f1db6-dbcc-483e-87e4-3c8afd65fdb6	332ddb5a-b7e6-484d-8db9-88a7e381f49a	e3263895-747b-4a74-8346-e573778408e8	false
> > select uuid, name from unit
> uuid					name
> d0c384e3-6800-4b3c-800d-1a37740aa189	source/0
> e3263895-747b-4a74-8346-e573778408e8	sink/0
> > select * from relation_unit_setting
> relation_unit_uuid			key		value
> e55c3a39-6309-4a6e-8f07-da8b3a2f553f	ingress-address	local-cloud:10.41.184.38@space:656b4a82-e28c-53d6-a014-f0dd53417eb6
> e55c3a39-6309-4a6e-8f07-da8b3a2f553f	token		yeah-boi
> ba4f1db6-dbcc-483e-87e4-3c8afd65fdb6	ingress-address	local-cloud:10.41.184.140@space:656b4a82-e28c-53d6-a014-f0dd53417eb6
> ```
> 
> I reverted the one api sha and did:
> 
> ```
> $ juju add-model seven; juju deploy juju-qa-dummy-source source --config  token=yeah-boi; juju deploy juju-qa-dummy-sink sink; juju relate source sink
> ```

I've already fixed that one here https://github.com/juju/juju/pull/19620! It got merged EoD yesterday so I didn't have time to rebase this one, but I did testing with it.

Aflynn50 commented:
/merge
